### PR TITLE
Refs #36760 - Reset candlepin key- and truststore

### PR DESCRIPTION
**Pull request by ekohl**
This takes a 2 step approach where the cached password is removed during migration (which ends up running during the RPM installation). The installer then handles replacing the stores when it really runs.

The cherry pick also includes module updates which are required to fix the issue.

(cherry picked from commit a45ac09dd9a5f039ab84612df6bd7f5cd0e5b960)

### CODE REVIEW

**Comment by ekohl** [comment]
I hadn't merged https://github.com/theforeman/puppet-certs/pull/431 yet
[test foreman-installer]
/packit build

**Comment by ekohl** [comment]
Hold off on this, I missed something:
```
[root@centos8-stream-katello-4-10 ~]# foreman-installer
2023-10-05 16:32:11 [NOTICE] [root] Loading installer configuration. This will take some time.
2023-10-05 16:32:15 [NOTICE] [root] Running installer with log based terminal output at level NOTICE.
2023-10-05 16:32:15 [NOTICE] [root] Use -l to set the terminal output log level to ERROR, WARN, NOTICE, INFO, or DEBUG. See --full-help for definitions.
2023-10-05 16:32:21 [NOTICE] [configure] Starting system configuration.
2023-10-05 16:32:28 [NOTICE] [configure] 250 configuration steps out of 1407 steps complete.
2023-10-05 16:32:29 [ERROR ] [configure] Failed to read keystore '/etc/candlepin/certs/keystore'
2023-10-05 16:32:29 [ERROR ] [configure] Failed to read keystore '/etc/candlepin/certs/truststore'
2023-10-05 16:32:30 [NOTICE] [configure] 500 configuration steps out of 1409 steps complete.
2023-10-05 16:32:31 [ERROR ] [configure] Execution of '/bin/keytool -import -v -noprompt -storetype pkcs12 -keystore /etc/candlepin/certs/truststore -alias artemis-client -file /etc/foreman/client_cert.pem -storepass:file /etc/pki/katello/truststore_password-file -J-Dcom.redhat.fips=false' returned 1: keytool error: java.io.IOException: keystore password was incorrect
2023-10-05 16:32:31 [ERROR ] [configure] java.io.IOException: keystore password was incorrect
2023-10-05 16:32:31 [ERROR ] [configure] at sun.security.pkcs12.PKCS12KeyStore.engineLoad(PKCS12KeyStore.java:2089)
2023-10-05 16:32:31 [ERROR ] [configure] at java.security.KeyStore.load(KeyStore.java:1445)
2023-10-05 16:32:31 [ERROR ] [configure] at sun.security.tools.keytool.Main.doCommands(Main.java:839)
2023-10-05 16:32:31 [ERROR ] [configure] at sun.security.tools.keytool.Main.run(Main.java:380)
2023-10-05 16:32:31 [ERROR ] [configure] at sun.security.tools.keytool.Main.main(Main.java:373)
2023-10-05 16:32:31 [ERROR ] [configure] Caused by: java.security.UnrecoverableKeyException: failed to decrypt safe contents entry: javax.crypto.BadPaddingException: Given final block not properly padded. Such issues can arise if a bad key is used during decryption.
2023-10-05 16:32:31 [ERROR ] [configure] ... 5 more
2023-10-05 16:32:31 [ERROR ] [configure] /Stage[main]/Certs::Candlepin/Truststore_certificate[/etc/candlepin/certs/truststore:artemis-client]/ensure: change from 'absent' to 'present' failed: Execution of '/bin/keytool -import -v -noprompt -storetype pkcs12 -keystore /etc/candlepin/certs/truststore -alias artemis-client -file /etc/foreman/client_cert.pem -storepass:file /etc/pki/katello/truststore_password-file -J-Dcom.redhat.fips=false' returned 1: keytool error: java.io.IOException: keystore password was incorrect
2023-10-05 16:32:31 [ERROR ] [configure] java.io.IOException: keystore password was incorrect
2023-10-05 16:32:31 [ERROR ] [configure] at sun.security.pkcs12.PKCS12KeyStore.engineLoad(PKCS12KeyStore.java:2089)
2023-10-05 16:32:31 [ERROR ] [configure] at java.security.KeyStore.load(KeyStore.java:1445)
2023-10-05 16:32:31 [ERROR ] [configure] at sun.security.tools.keytool.Main.doCommands(Main.java:839)
2023-10-05 16:32:31 [ERROR ] [configure] at sun.security.tools.keytool.Main.run(Main.java:380)
2023-10-05 16:32:31 [ERROR ] [configure] at sun.security.tools.keytool.Main.main(Main.java:373)
2023-10-05 16:32:31 [ERROR ] [configure] Caused by: java.security.UnrecoverableKeyException: failed to decrypt safe contents entry: javax.crypto.BadPaddingException: Given final block not properly padded. Such issues can arise if a bad key is used during decryption.
2023-10-05 16:32:31 [ERROR ] [configure] ... 5 more
2023-10-05 16:32:31 [ERROR ] [configure] Failed to add certificate to keystore: Execution of '/bin/keytool -importkeystore -noprompt -srckeystore /tmp/temp_keystore20231005-45997-149x4ud -srcstorepass:file /etc/pki/katello/keystore_password-file -destkeystore /etc/candlepin/certs/keystore -deststorepass:file /etc/pki/katello/keystore_password-file -srcalias tomcat -destalias tomcat -J-Dcom.redhat.fips=false' returned 1: Importing keystore /tmp/temp_keystore20231005-45997-149x4ud to /etc/candlepin/certs/keystore...
2023-10-05 16:32:31 [ERROR ] [configure] keytool error: java.io.IOException: keystore password was incorrect
2023-10-05 16:32:32 [ERROR ] [configure] Execution of '/bin/keytool -import -v -noprompt -storetype pkcs12 -keystore /etc/candlepin/certs/truststore -alias candlepin-ca -file /etc/candlepin/certs/candlepin-ca.crt -storepass:file /etc/pki/katello/truststore_password-file -J-Dcom.redhat.fips=false' returned 1: keytool error: java.io.IOException: keystore password was incorrect
2023-10-05 16:32:32 [ERROR ] [configure] java.io.IOException: keystore password was incorrect
2023-10-05 16:32:32 [ERROR ] [configure] at sun.security.pkcs12.PKCS12KeyStore.engineLoad(PKCS12KeyStore.java:2089)
2023-10-05 16:32:32 [ERROR ] [configure] at java.security.KeyStore.load(KeyStore.java:1445)
2023-10-05 16:32:32 [ERROR ] [configure] at sun.security.tools.keytool.Main.doCommands(Main.java:839)
2023-10-05 16:32:32 [ERROR ] [configure] at sun.security.tools.keytool.Main.run(Main.java:380)
2023-10-05 16:32:32 [ERROR ] [configure] at sun.security.tools.keytool.Main.main(Main.java:373)
2023-10-05 16:32:32 [ERROR ] [configure] Caused by: java.security.UnrecoverableKeyException: failed to decrypt safe contents entry: javax.crypto.BadPaddingException: Given final block not properly padded. Such issues can arise if a bad key is used during decryption.
2023-10-05 16:32:32 [ERROR ] [configure] ... 5 more
2023-10-05 16:32:32 [ERROR ] [configure] /Stage[main]/Certs::Candlepin/Truststore_certificate[/etc/candlepin/certs/truststore:candlepin-ca]/ensure: change from 'absent' to 'present' failed: Execution of '/bin/keytool -import -v -noprompt -storetype pkcs12 -keystore /etc/candlepin/certs/truststore -alias candlepin-ca -file /etc/candlepin/certs/candlepin-ca.crt -storepass:file /etc/pki/katello/truststore_password-file -J-Dcom.redhat.fips=false' returned 1: keytool error: java.io.IOException: keystore password was incorrect
2023-10-05 16:32:32 [ERROR ] [configure] java.io.IOException: keystore password was incorrect
2023-10-05 16:32:32 [ERROR ] [configure] at sun.security.pkcs12.PKCS12KeyStore.engineLoad(PKCS12KeyStore.java:2089)
2023-10-05 16:32:32 [ERROR ] [configure] at java.security.KeyStore.load(KeyStore.java:1445)
2023-10-05 16:32:32 [ERROR ] [configure] at sun.security.tools.keytool.Main.doCommands(Main.java:839)
2023-10-05 16:32:32 [ERROR ] [configure] at sun.security.tools.keytool.Main.run(Main.java:380)
2023-10-05 16:32:32 [ERROR ] [configure] at sun.security.tools.keytool.Main.main(Main.java:373)
2023-10-05 16:32:32 [ERROR ] [configure] Caused by: java.security.UnrecoverableKeyException: failed to decrypt safe contents entry: javax.crypto.BadPaddingException: Given final block not properly padded. Such issues can arise if a bad key is used during decryption.
2023-10-05 16:32:32 [ERROR ] [configure] ... 5 more
2023-10-05 16:32:32 [NOTICE] [configure] 750 configuration steps out of 1413 steps complete.
2023-10-05 16:32:32 [NOTICE] [configure] 1000 configuration steps out of 1418 steps complete.
2023-10-05 16:32:39 [NOTICE] [configure] 1250 configuration steps out of 1418 steps complete.
2023-10-05 16:32:45 [NOTICE] [configure] System configuration has finished.

Error 1: Puppet Truststore_certificate resource '/etc/candlepin/certs/truststore:artemis-client' failed. Logs:
  /Stage[main]/Certs::Candlepin/Truststore_certificate[/etc/candlepin/certs/truststore:artemis-client]
    Adding autorequire relationship with File[/etc/pki/katello/truststore_password-file]
    Adding autorequire relationship with File[/etc/foreman/client_cert.pem]
    Adding autonotify relationship with File[/etc/candlepin/certs/truststore]
    Starting to evaluate the resource (577 of 1409)
    Evaluated in 0.34 seconds
  /Stage[main]/Certs::Candlepin/Truststore_certificate[/etc/candlepin/certs/truststore:artemis-client]/ensure
    change from 'absent' to 'present' failed: Execution of '/bin/keytool -import -v -noprompt -storetype pkcs12 -keystore /etc/candlepin/certs/truststore -alias artemis-client -file /etc/foreman/client_cert.pem -storepass:file /etc/pki/katello/truststore_password-file -J-Dcom.redhat.fips=false' returned 1: keytool error: java.io.IOException: keystore password was incorrect
java.io.IOException: keystore password was incorrect
	at sun.security.pkcs12.PKCS12KeyStore.engineLoad(PKCS12KeyStore.java:2089)
	at java.security.KeyStore.load(KeyStore.java:1445)
	at sun.security.tools.keytool.Main.doCommands(Main.java:839)
	at sun.security.tools.keytool.Main.run(Main.java:380)
	at sun.security.tools.keytool.Main.main(Main.java:373)
Caused by: java.security.UnrecoverableKeyException: failed to decrypt safe contents entry: javax.crypto.BadPaddingException: Given final block not properly padded. Such issues can arise if a bad key is used during decryption.
	... 5 more
Error 2: Puppet Truststore_certificate resource '/etc/candlepin/certs/truststore:candlepin-ca' failed. Logs:
  /Stage[main]/Certs::Candlepin/Truststore_certificate[/etc/candlepin/certs/truststore:candlepin-ca]
    Adding autorequire relationship with File[/etc/pki/katello/truststore_password-file]
    Adding autorequire relationship with File[/etc/candlepin/certs/candlepin-ca.crt]
    Adding autonotify relationship with File[/etc/candlepin/certs/truststore]
    Starting to evaluate the resource (598 of 1409)
    Evaluated in 0.33 seconds
  /Stage[main]/Certs::Candlepin/Truststore_certificate[/etc/candlepin/certs/truststore:candlepin-ca]/ensure
    change from 'absent' to 'present' failed: Execution of '/bin/keytool -import -v -noprompt -storetype pkcs12 -keystore /etc/candlepin/certs/truststore -alias candlepin-ca -file /etc/candlepin/certs/candlepin-ca.crt -storepass:file /etc/pki/katello/truststore_password-file -J-Dcom.redhat.fips=false' returned 1: keytool error: java.io.IOException: keystore password was incorrect
java.io.IOException: keystore password was incorrect
	at sun.security.pkcs12.PKCS12KeyStore.engineLoad(PKCS12KeyStore.java:2089)
	at java.security.KeyStore.load(KeyStore.java:1445)
	at sun.security.tools.keytool.Main.doCommands(Main.java:839)
	at sun.security.tools.keytool.Main.run(Main.java:380)
	at sun.security.tools.keytool.Main.main(Main.java:373)
Caused by: java.security.UnrecoverableKeyException: failed to decrypt safe contents entry: javax.crypto.BadPaddingException: Given final block not properly padded. Such issues can arise if a bad key is used during decryption.
	... 5 more

2 errors were detected.
Please address the errors and re-run the installer to ensure the system is properly configured.
Failing to do so is likely to result in broken functionality.

The full log is at /var/log/foreman-installer/katello.log
```

I'm not sure why I didn't touch the truststore and only did the keystore.

**Comment by ekohl** [comment]
https://github.com/theforeman/puppet-certs/pull/432 should address that.

**Comment by ekohl** [comment]
Now verified on a Katello 4.10 box using packit.

File: Puppetfile [hunk 1]
```diff
@@ -17,5 +17,5 @@ mod 'theforeman/foreman',              '~> 23.1.0'
 mod 'theforeman/foreman_proxy',        '~> 25.1.0'
 mod 'theforeman/puppet',               '~> 18.0.0'
 mod 'katello/foreman_proxy_content',   '~> 25.0.0'
-mod 'katello/certs',                   '~> 16.0.2'
+mod 'katello/certs',                   '~> 16.1.0'
```
Comment (evgeni) [inline Puppetfile]: should this enforce `>= 16.1.1` given 16.1.0 is "broken"?

File: Puppetfile.lock [hunk 2]
```diff
@@ -1,12 +1,12 @@
 FORGE
   remote: https://forgeapi.puppet.com/
   specs:
-    katello-candlepin (14.1.1)
+    katello-candlepin (14.1.2)
```
Comment (evgeni) [inline Puppetfile.lock:4]: same, should `>= 14.1.2` be enforced, as otherwise this does not fix the issue?

File: Puppetfile (hunk 1 above)
Comment (ekohl) [reply Puppetfile]: Technically it should be sufficient to enforce it via the lock file, but I guess it doesn't hurt.

**Review by evgeni — APPROVED** [review]

**Pull request merged**